atef/bin/check.py

```diff
@@ -0,0 +1,254 @@
+            continue
+
+        try:
+            search_result = client[dev_name]
+        except KeyError:
+            ex = MissingHappiDeviceError(
+                f"Device {dev_name} not in happi database; skipping"
+            )
+            ex.dev_name = dev_name
+            ex.dev_config = cast(DeviceConfiguration, dev_config)
+            yield ex
+            continue
+
+        try:
+            dev = search_result.get()
+        except Exception as ex:
+            logger.debug(
+                "Failed to instantiate device %r",
+                dev_name,
+                exc_info=True,
+            )
+            load_ex = HappiLoadError(
+                f"Device {dev_name} invalid in happi database; "
+                f"{ex.__class__.__name__}: {ex}"
+            )
+            load_ex.dev_name = dev_name
+            load_ex.dev_config = cast(DeviceConfiguration, dev_config)
+            yield load_ex
+            continue
+
+        yield DeviceAndConfig(dev, dev_config)
+
+
+def log_results(
+    device: ophyd.Device,
+    severity: Severity,
+    config: Union[PVConfiguration, DeviceConfiguration],
+    results: List[Result],
+    *,
+    severity_to_log_level: Optional[Dict[Severity, int]] = None,
+):
+    """Log check results to the module logger."""
+    severity_to_log_level = severity_to_log_level or default_severity_to_log_level
+
+    passed_or_failed = "passed" if severity <= Severity.warning else "FAILED"
+    logger.info(
+        "Device %s (%s) %s with severity %s",
+        device.name,
+        config.description or "no description",
+        passed_or_failed,
+        severity.name,
+    )
+    for result in results:
+        log_level = severity_to_log_level[result.severity]
+        if not logger.isEnabledFor(log_level):
+            continue
+        logger.log(log_level, result.reason)
+
+
+def log_results_rich(
+    console: rich.console.Console,
+    device: ophyd.Device,
+    severity: Severity,
+    config: Union[PVConfiguration, DeviceConfiguration],
+    results: List[Result],
+    *,
+    severity_to_rich: Optional[Dict[Severity, str]] = None,
+    verbose: int = 0,
+):
+    """Log check results to the module logger."""
+    severity_to_rich = severity_to_rich or default_severity_to_rich
+
+    desc = f" ({config.description}) " if config.description else ""
+
+    tree = rich.tree.Tree(f"{severity_to_rich[severity]} [default]{device.name}{desc}")
+    for result in results:
+        if result.severity > Severity.success or verbose > 0:
+            tree.add(
+                f"{severity_to_rich[result.severity]}[default]: {result.reason}"
+            )
+    console.print(tree)
+
+
+def main(
+    filename: str,
+    filtered_devices: Optional[Sequence[str]] = None,
+    verbose: int = 0,
+    *,
+    cleanup: bool = True
+):
+    serialized_config = yaml.safe_load(open(filename))
+    config = apischema.deserialize(ConfigurationFile, serialized_config)
```

Comment:
Somehow I missed that this was possible- I'm giving apischema another look now. Very cool.